packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts

```diff
@@ -0,0 +1,149 @@
+      const resource = node as Resource;
+      if (this.exclude.length !== 0 && this.exclude.indexOf(resource.resourceType) !== -1) {
+        return;
+      }
+      if (this.include.length !== 0 && this.include.indexOf(resource.resourceType) === -1) {
+        return;
+      }
+      if (this.hasPrecedence(node.aspects)) {
+        this.applyTag(resource);
+      }
+    }
+  }
+
+  protected abstract applyTag(resource: Resource): void;
+
+  private hasPrecedence(aspects: IAspect[]): boolean {
+    const tagAspects = aspects.filter( a => a !== this && a.type === this.type);
+    const collisions = tagAspects.filter( a => (a as TagBaseAspect).key === this.key);
+    for (const collidingAspect of collisions) {
+      const collision = collidingAspect as TagBaseAspect;
+      // if the parent is not defined the apsect hasn't been visited and is
+      // local to the current node; thus the collision has precedence because it
+      // is later (e.g. last one in wins) in the array
+      if (collision.parent === undefined) {
+        return false;
+      }
+      // the aspect with the most ancestors is closest to this node in the tree
+      if (collision.parent.ancestors().length > this.parent!.ancestors().length) {
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+export class TagAspect extends TagBaseAspect {
+
+  constructor(key: string, value: string, props: TagProperties = {}) {
+    super(key, {value, ...props});
+    if (this.value === undefined) {
+      throw new Error('Tag must have a value');
+    }
+  }
+
+  protected applyTag(resource: Resource) {
+    removeTag(this, resource);
+    setTag(this, resource);
+  }
+}
+
+export class RemoveTag extends TagBaseAspect {
+
+  constructor(key: string, props: TagProperties = {}) {
+    super(key, props);
+  }
+
+  protected applyTag(resource: Resource) {
+    return removeTag(this, resource);
+  }
+}
+
+function removeTag(tag: TagBaseAspect, resource: Resource): void {
+  const key = tag.key;
+  switch (resource.tagType) {
+    case TagType.Standard:
+    case TagType.AutoScalingGroup: {
+      if (resource.properties.tags === undefined) { break; }
+      resource.properties.tags = resource.properties.tags.filter( (t: any ) => t.key !== key);
+      break;
+    }
+    case TagType.Map: {
+      if (resource.properties.tags === undefined) { break; }
+      delete resource.properties.tags[key];
+      break;
+    }
+  }
+}
+
+function setTag(tag: TagBaseAspect, resource: Resource) {
+  switch (resource.tagType) {
+    case TagType.Standard: {
+      resource.properties.tags = resource.properties.tags || [];
+      resource.properties.tags.push({ key: tag.key, value: tag.value });
+      break;
+    }
+    case TagType.AutoScalingGroup: {
+      resource.properties.tags = resource.properties.tags || [];
+      resource.properties.tags.push({ key: tag.key, value: tag.value, propagateAtLaunch: tag.propagate !== 0});
```

Comment:
I'm totally fine with `propagateAtLaunch` being a property of `Tag` (maybe called `applyToLaunchedInstances`), which only applies in case the tag is applied to AutoScalingGroups.

This mechanism of "having enough propagates left once the tag gets applied" seems hard to understand.